text/0000-unsafe-statics.md

```diff
@@ -0,0 +1,98 @@
+- Feature Name: unsafe_static
+- Start Date: 2020-05-29
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+Replace static muts with unsafe statics, which are error prone.
+
+# Motivation
+[motivation]: #motivation
+
+Since before 1.0, Rust has had a feature called `static mut`, which allows users to define statics
+to which mutable references can be taken. Actually taking a mutable reference is unsafe. However,
+this feature is incredibly footgunny and almost impossible to use correctly, since creating a
+mutable reference to shared memory is undefined behavior, even if you never dereference that
+reference.
+
+Unsafe statics would be a better way to get the same effect, by requiring any mutable access to
+still pass through an interior mutability primitive like `UnsafeCell`.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+statics can now be declared `unsafe`:
+
+```rust
+unsafe static X: i32 = 0;
+```
+
+Statics that are declared unsafe are different from other statics in the following ways:
+
+1. All accesses to them are unsafe.
+2. The `Sync` check is disabled for unsafe statics. They are not required to implement `Sync`.
+
+Unlike static muts, it is not possible to take a mutable reference to an unsafe static. However
+users can achieve the effect of a static mut using an `UnsafeCell`:
+
+```rust
+unsafe static X: UnsafeCell<i32> = UnsafeCell::new(0);
+// safer equivalent of:
+static mut X: i32 = 0;
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## Use cases
+
+Users using this like a static mut are responsible for guaranteeing that they never construct a
+mutable reference to the inner value in an unsynchronized manner. This is essentially the same as
+static mut's use case, but more sound in the face of the actual current definition off undefined
+behavior.
+
+Users could also achieve the effect of an unsafe thread local, which has no synchronization, but
+which they guarantee they only access from one thread. For example, they could use it with `Cell`
+and `RefCell`. (They can do this with static mut too, but with this API they are guaranteed never to
+get a totally unsynchronized mutable reference, even if they are responsible for guaranteeing thread
+locality.)
+
+Another interesting use case would be to use a totally Sync, shared primitive, but introduce new
+invariants which must be upheld. In other words, unsafe statics could behave like unsafe traits in
+creating new invariant abstraction points. It's unclear what use cases this would have, but it seems
+like something which could be compelling.
```

Comment:
Oh I see, so that would be using an `unsafe static` for "communication" between a library and its environment, and make that communication subject to extra constraints? Yes that sounds potentially interesting indeed.

(I struggled at first to understand this paragraph, it would help if you said explicitly that a hypothetical crate would make such an `unsafe static` part of its public API.)